horovod/common/controller.cc

```diff
@@ -628,53 +619,45 @@ void Controller::CoordinateCacheAndState(CacheCoordinator& cache_coordinator) {
                                (Request::RequestType)response.response_type());
     }
 
-    // End negotation phase for synced cache hit set entries.
+    // End negotiation phase for synced cache hit set entries.
     for (auto bit : cache_coordinator.cache_hits()) {
       auto& response = response_cache_.peek_response(bit);
       timeline_.NegotiateEnd(response.tensor_names()[0]);
     }
   }
 }
 
-ResponseList Controller::FuseResponses(std::deque<Response>& responses,
-                                       bool joined) {
+ResponseList Controller::FuseResponses(std::deque<Response>& responses) {
   ResponseList response_list;
   while (!responses.empty()) {
 
     auto response = responses.front();
     assert(response.tensor_names().size() == 1);
     responses.pop_front();
     int64_t tensor_size = 0;
-    DataType dtype;
     if (response.response_type() == Response::ResponseType::ALLREDUCE ||
         response.response_type() == Response::ResponseType::ADASUM) {
       // Attempt to add more responses to this fused response.
 
+      tensor_size = response.tensor_sizes()[0];
       std::deque<Response> skipped_responses;
       int64_t skipped_size = 0;
       while (!responses.empty()) {
         auto new_response = responses.front();
         assert(new_response.tensor_names().size() == 1);
 
-        std::vector<TensorTableEntry> entries_for_join;
-        if (joined) {
-          tensor_queue_.GetTensorEntriesFromResponse(response, entries_for_join,
-                                                     joined);
-        }
-
-        const auto& new_entry =
-            joined
-                ? entries_for_join[0]
-                : tensor_queue_.GetTensorEntry(new_response.tensor_names()[0]);
-        int64_t new_tensor_size = new_entry.tensor->size();
-
+        int64_t new_tensor_size = new_response.tensor_sizes().empty()
```

Comment:
Nice catch @kit1980! If the tests pass with this change, I suggest modifying this so that rather than creating and setting the `new_tensor_size` variable out here with this check on `new_response.tensor_sizes().empty()`, just use `new_response.tensor_sizes()[0]` directly in place of `new_tensor_size` in code below. The first check on `response.response_type() == new_response.response_type()` in the if statement should guard against attempting to access the size for non allreduce messages.

Reply:
Hi @romerojosh. The tests passed with the current code in this PR.
I tried your suggestion, but then the tests fail after error cases, like duplicate name tests, (but not every time...)

Reply:
Yes, you are right. The segfault in the updated version occurs in the else case where `new_tensor_size` is being used (which becomes problematic if replaced with `new_response.tensor_sizes()[0]` directly):
https://github.com/horovod/horovod/blob/f042aeca6b9e9a273b43fd15fd22e7dac970827e/horovod/common/controller.cc#L672

The existing solution is fine with me, but @tgaddair, maybe you can also chime in on this?

Reply:
I think this approach is fine as well.